### ISSUE TITLE
pybullet.changeVisualShape broken when applying texture mesh and binding an external renderer

### ISSUE THREAD

**Issue by juancamilog** [ikalevatykh/pybullet_rendering#8]
The following code runs fine with egl or tiny renderer

    import pybullet as p
    import pybullet_data
    import numpy as np
    from pybullet_rendering import RenderingPlugin
    from pybullet_rendering.render.panda3d import Renderer

    def create_plane_mesh(resx=3, resy=3):
        verts = np.stack(
            np.meshgrid(np.linspace(-0.5, 0.5, resx),
                        np.linspace(-0.5, 0.5, resy))).reshape(2, -1).T
        verts = np.concatenate([verts, np.zeros((len(verts), 1))], -1)
        tris = []
        for i in range(resx - 1):
            for j in range(resy - 1):
                ii = i + resx * j
                tris += [ii, ii + 1, ii + resx]
                tris += [ii + resx, ii + 1, ii + resx + 1]
        uvs = np.stack(
            np.meshgrid(np.linspace(0, 1, resx),
                        np.linspace(0, 1, resy))).reshape(2, -1).T
        normals = np.repeat([[0, 0, 1.0]], len(verts), 0)
        return verts, tris, uvs, normals

    physicsClientId = p.connect(p.GUI)#or p.DIRECT for non-graphical version

    # bind renderer to pybullet
    # renderer = Renderer(MSAA_samples=4)
    # plugin = RenderingPlugin(physicsClientId, renderer)

    # create texture image
    x, y = np.meshgrid(np.linspace(-1,1, 128), np.linspace(-1,1, 128))
    img = (72*(np.stack([np.cos(16*x), np.cos(16*y), np.cos(16*(x+y))])+2)).astype(np.uint8).transpose(1,2,0)
    from PIL import Image
    img = Image.fromarray(img)
    fname = 'panda3d_texture_test.png'
    img.save(fname)

    # create plane mesh
    verts, tris, uvs, normals = create_plane_mesh(128, 128)
    verts[:, 0] += 0.05*np.cos(8*x.flatten())
    verts[:, 1] += 0.05*np.cos(8*y.flatten())
    verts[:, 2] += 0.1*np.cos(8*(x+y).flatten())
    planeColl = p.createCollisionShape(shapeType=p.GEOM_MESH,
                                       meshScale=[1, 1, 1],
                                       vertices=verts,
                                       indices=tris,
                                       physicsClientId=physicsClientId)

    planeVis = p.createVisualShape(shapeType=p.GEOM_MESH,
                                   rgbaColor=[1, 1, 1, 1],
                                   specularColor=[0.4, .4, 0],
                                   meshScale=[1, 1, 1],
                                   vertices=verts,
                                   indices=tris,
                                   uvs=uvs,
                                   normals=normals,
                                   physicsClientId=physicsClientId)

    planeId = p.createMultiBody(0,
                                baseCollisionShapeIndex=planeColl,
                                baseVisualShapeIndex=planeVis,
                                physicsClientId=physicsClientId)

    # apply texture to plane
    textureId = p.loadTexture(fname, physicsClientId=physicsClientId)
    p.changeVisualShape(planeId,
                        -1,
                        textureUniqueId=textureId,
                        rgbaColor=[1, 1, 1, 1],
                        specularColor=[1, 1, 1, 1],
                        physicsClientId=physicsClientId)

    # p.disconnect()

Resulting in the following image
![image](https://user-images.githubusercontent.com/2013726/95141403-16b18980-073f-11eb-8457-d17e8c7b7f19.png)

Binding a panda3d or pyrender renderer, the code above crashes when calling `p.changeVisualShape` with the following error message

    terminate called after throwing an instance of 'std::out_of_range'
    what():  map::at

I also tried populating the  link arrays for the `p.createMultiBody` method and callling `p.changeVisualShape` with the joint index equal to 0 instead of -1, but i got the same result.

TL; DR changeVisualShape is broken with manually created instances of multiBody objects, when the binding the external renderers from this repo. Do you have nay suggestions on how to fix this?

**Comment by ikalevatykh** [comment]
Hi @juancamilog, procedural generated meshes not yet supported.

**Comment by juancamilog** [comment]
There is a workaround by writing mesh files to disk and loading them with the pybullet API

    import pybullet as p
    import pybullet_data
    import numpy as np
    import trimesh
    import textwrap
    from pybullet_rendering import RenderingPlugin
    from pybullet_rendering.render.panda3d import Renderer


    def create_plane_mesh(resx=3, resy=3):
        verts = np.stack(
            np.meshgrid(np.linspace(-0.5, 0.5, resx),
                        np.linspace(-0.5, 0.5, resy))).reshape(2, -1).T
        verts = np.concatenate([verts, np.zeros((len(verts), 1))], -1)
        tris = []
        for i in range(resx - 1):
            for j in range(resy - 1):
                ii = i + resx * j
                tris += [ii, ii + 1, ii + resx]
                tris += [ii + resx, ii + 1, ii + resx + 1]
        uvs = np.stack(
            np.meshgrid(np.linspace(0, 1, resx),
                        np.linspace(0, 1, resy))).reshape(2, -1).T
        normals = np.repeat([[0, 0, 1.0]], len(verts), 0)
        return verts, tris, uvs, normals


    def write_obj(verts, tris, uvs, normals, img, path):
        mesh = trimesh.Trimesh(vertices=verts, faces=np.array(tris).reshape(-1,3), vertex_normals=normals, process=False)
        mesh.visual = trimesh.visual.TextureVisuals(uv=uvs, image=img)
        with open(path, 'w') as f:
            f.write(trimesh.exchange.obj.export_obj(mesh))

    physicsClientId = p.connect(p.GUI)#or p.DIRECT for non-graphical version

    # bind renderer to pybullet
    renderer = Renderer(MSAA_samples=4)
    plugin = RenderingPlugin(physicsClientId, renderer)

    # create texture image
    x, y = np.meshgrid(np.linspace(-1,1, 128), np.linspace(-1,1, 128))
    img = (72*(np.stack([np.cos(16*x), np.cos(16*y), np.cos(16*(x+y))])+2)).astype(np.uint8).transpose(1,2,0)
    from PIL import Image
    img = Image.fromarray(img)
    fname = 'panda3d_texture_test.png'
    img.save(fname)

    # create plane mesh
    verts, tris, uvs, normals = create_plane_mesh(128, 128)
    verts[:, 0] += 0.05*np.cos(8*x.flatten())
    verts[:, 1] += 0.05*np.cos(8*y.flatten())
    verts[:, 2] += 0.1*np.cos(8*(x+y).flatten())

    # load visual mesh
    write_obj(verts, tris, uvs, normals, img, 'test_mesh.obj')

    planeColl = p.createCollisionShape(shapeType=p.GEOM_MESH,
                                       fileName='test_mesh.obj',
                                       physicsClientId=physicsClientId)

    planeVis = p.createVisualShape(shapeType=p.GEOM_MESH,
                                   fileName='test_mesh.obj',
                                   rgbaColor=[1, 1, 1, 1],
                                   specularColor=[0.4, .4, 0],
                                   physicsClientId=physicsClientId)

    planeId = p.createMultiBody(0,
                                baseCollisionShapeIndex=planeColl,
                                baseVisualShapeIndex=planeVis,
                                physicsClientId=physicsClientId)

    # apply texture to plane
    textureId = p.loadTexture(fname, physicsClientId=physicsClientId)
    p.changeVisualShape(planeVis,
                        -1,
                        textureUniqueId=textureId,
                        rgbaColor=[1, 1, 1, 1],
                        specularColor=[1, 1, 1, 1],
                        physicsClientId=physicsClientId)

I'm not entirely sure if this supports non-concave meshes

**Comment by juancamilog** [comment]
Created PR #9  to update panda3d and pyrender renderers to load textures properly when they exist.

**Comment by ikalevatykh** [comment]
Hi @juancamilog , sorry for late response.

Thank you for reporting, this should work with the latest version.
